From 8ab03748dee2ef57c4e7e25c52ff7d171a451228 Mon Sep 17 00:00:00 2001
From: tropicrainforest <tropicrainforest27017@gmail.com>
Date: Sat, 1 Aug 2026 18:10:50 +0200
Subject: [PATCH 3/5] mozjs: gate wrapper-comparison operators ElementType
 behind DefineComparisonOps for MSVC 14.4x

The templated ==/!= operators in js/ComparisonOperators.h name the
wrapper's nested ElementType unconditionally. Under C++20, MSVC 14.4x
instantiates that reference even for wrappers that do not define
ElementType (DefineComparisonOps == false), which fails to compile.
Hier we introduce a ComparableElementType trait keyed on DefineComparisonOps<W>
so ElementType is only named when the wrapper opts in; otherwise it falls
back to the wrapper type itself.
---
 .../mozjs/include/js/ComparisonOperators.h    | 33 ++++++++++++-------
 1 file changed, 21 insertions(+), 12 deletions(-)

diff --git a/src/third_party/mozjs/include/js/ComparisonOperators.h b/src/third_party/mozjs/include/js/ComparisonOperators.h
index c7f03fa4c..423150c7d 100644
--- a/src/third_party/mozjs/include/js/ComparisonOperators.h
+++ b/src/third_party/mozjs/include/js/ComparisonOperators.h
@@ -49,6 +49,15 @@ namespace detail {
 template <typename T>
 struct DefineComparisonOps : std::false_type {};
 
+template <typename W, bool = DefineComparisonOps<W>::value>
+struct ComparableElementType {
+  using Type = W;
+};
+template <typename W>
+struct ComparableElementType<W, true> {
+  using Type = typename W::ElementType;
+};
+
 // Define functions for the core equality operations, that the actual operators
 // can all invoke.
 
@@ -111,25 +120,25 @@ operator!=(const W& wrapper, const OW& other) {
 // Comparisons between a wrapper and its unwrapped element type.
 template <typename W>
 inline typename std::enable_if_t<DefineComparisonOps<W>::value, bool>
-operator==(const W& wrapper, const typename W::ElementType& value) {
+operator==(const W& wrapper, const typename ComparableElementType<W>::Type& value) {
   return WrapperEqualsUnwrapped(wrapper, value);
 }
 
 template <typename W>
 inline typename std::enable_if_t<DefineComparisonOps<W>::value, bool>
-operator!=(const W& wrapper, const typename W::ElementType& value) {
+operator!=(const W& wrapper, const typename ComparableElementType<W>::Type& value) {
   return !WrapperEqualsUnwrapped(wrapper, value);
 }
 
 template <typename W>
 inline typename std::enable_if_t<DefineComparisonOps<W>::value, bool>
-operator==(const typename W::ElementType& value, const W& wrapper) {
+operator==(const typename ComparableElementType<W>::Type& value, const W& wrapper) {
   return WrapperEqualsUnwrapped(wrapper, value);
 }
 
 template <typename W>
 inline typename std::enable_if_t<DefineComparisonOps<W>::value, bool>
-operator!=(const typename W::ElementType& value, const W& wrapper) {
+operator!=(const typename ComparableElementType<W>::Type& value, const W& wrapper) {
   return !WrapperEqualsUnwrapped(wrapper, value);
 }
 
@@ -137,36 +146,36 @@ operator!=(const typename W::ElementType& value, const W& wrapper) {
 // and a const element pointer.
 template <typename W>
 inline typename std::enable_if_t<DefineComparisonOps<W>::value &&
-                                     std::is_pointer_v<typename W::ElementType>,
+                                     std::is_pointer_v<typename ComparableElementType<W>::Type>,
                                  bool>
 operator==(const W& wrapper,
-           const typename std::remove_pointer_t<typename W::ElementType>* ptr) {
+           const typename std::remove_pointer_t<typename ComparableElementType<W>::Type>* ptr) {
   return WrapperEqualsPointer(wrapper, ptr);
 }
 
 template <typename W>
 inline typename std::enable_if_t<DefineComparisonOps<W>::value &&
-                                     std::is_pointer_v<typename W::ElementType>,
+                                     std::is_pointer_v<typename ComparableElementType<W>::Type>,
                                  bool>
 operator!=(const W& wrapper,
-           const typename std::remove_pointer_t<typename W::ElementType>* ptr) {
+           const typename std::remove_pointer_t<typename ComparableElementType<W>::Type>* ptr) {
   return !WrapperEqualsPointer(wrapper, ptr);
 }
 
 template <typename W>
 inline typename std::enable_if_t<DefineComparisonOps<W>::value &&
-                                     std::is_pointer_v<typename W::ElementType>,
+                                     std::is_pointer_v<typename ComparableElementType<W>::Type>,
                                  bool>
-operator==(const typename std::remove_pointer_t<typename W::ElementType>* ptr,
+operator==(const typename std::remove_pointer_t<typename ComparableElementType<W>::Type>* ptr,
            const W& wrapper) {
   return WrapperEqualsPointer(wrapper, ptr);
 }
 
 template <typename W>
 inline typename std::enable_if_t<DefineComparisonOps<W>::value &&
-                                     std::is_pointer_v<typename W::ElementType>,
+                                     std::is_pointer_v<typename ComparableElementType<W>::Type>,
                                  bool>
-operator!=(const typename std::remove_pointer_t<typename W::ElementType>* ptr,
+operator!=(const typename std::remove_pointer_t<typename ComparableElementType<W>::Type>* ptr,
            const W& wrapper) {
   return !WrapperEqualsPointer(wrapper, ptr);
 }
